File: packages/agent_security/android/src/main/kotlin/com/chapter2/agentkit/models/WorldIdSignerBinding.kt
package com.chapter2.agentkit.models

public data class WorldIdSignerBinding(
    val signerAddress: String,
    val nullifierHash: String,
    val credentialType: WorldIdCredentialType,
    val boundAtTimestamp: Long = System.currentTimeMillis(),
    var lastActiveTimestamp: Long = System.currentTimeMillis(),
    var isRevoked: Boolean = false
) {
    var expiresAtTimestamp: Long = lastActiveTimestamp + SELFIE_INACTIVITY_WINDOW_MS

    fun isValid(currentTimestamp: Long = System.currentTimeMillis()): Boolean {
        if (isRevoked) return false
        return currentTimestamp < expiresAtTimestamp
    }

    fun touchActivity(currentTimestamp: Long = System.currentTimeMillis()) {
        this.lastActiveTimestamp = currentTimestamp
        this.expiresAtTimestamp = currentTimestamp + SELFIE_INACTIVITY_WINDOW_MS
    }

    companion object {
        const val SELFIE_INACTIVITY_WINDOW_MS: Long = 90L * 24 * 60 * 60 * 1000
    }
}

File: native_security/agentkit/ios/Sources/AgentKitSecurity/Models/WorldIdSignerBinding.swift
import Foundation

public struct WorldIdSignerBinding: Equatable, Codable, Sendable {
    public static let selfieInactivityWindowSeconds: TimeInterval = 90 * 24 * 3600

    public let signerAddress: String
    public let nullifierHash: String
    public let credentialType: WorldIdCredentialType
    public let boundAt: Date
    public var lastActiveAt: Date
    public var expiresAt: Date
    public var isRevoked: Bool

    public init(
        signerAddress: String,
        nullifierHash: String,
        credentialType: WorldIdCredentialType,
        boundAt: Date = Date(),
        lastActiveAt: Date = Date(),
        isRevoked: Bool = false
    ) {
        self.signerAddress = signerAddress
        self.nullifierHash = nullifierHash
        self.credentialType = credentialType
        self.boundAt = boundAt
        self.lastActiveAt = lastActiveAt
        self.expiresAt = lastActiveAt.addingTimeInterval(Self.selfieInactivityWindowSeconds)
        self.isRevoked = isRevoked
    }

    public func isValid(at currentDate: Date = Date()) -> Bool {
        guard !isRevoked else { return false }
        return currentDate < expiresAt
    }

    public mutating func touchActivity(at currentDate: Date = Date()) {
        self.lastActiveAt = currentDate
        self.expiresAt = currentDate.addingTimeInterval(Self.selfieInactivityWindowSeconds)
    }
}
